File: java-user-management-service/src/main/java/com/ghcp/starter/user_management/service/UserService.java
package com.ghcp.starter.user_management.service;

import com.ghcp.starter.user_management.dto.UserRegistrationRequest;
import com.ghcp.starter.user_management.dto.UserResponse;
import com.ghcp.starter.user_management.entity.User;
import com.ghcp.starter.user_management.entity.UserProfile;
import com.ghcp.starter.user_management.repository.UserRepository;
import com.ghcp.starter.user_management.repository.UserProfileRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import java.time.LocalDateTime;
import java.util.UUID;

@Service
public class UserService {
    @Autowired
    private UserRepository userRepository;
    @Autowired
    private UserProfileRepository userProfileRepository;
    @Autowired
    private PasswordEncoder passwordEncoder;

    public UserResponse registerUser(UserRegistrationRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Email already exists");
        }
        User user = new User();
        user.setId(UUID.randomUUID());
        user.setEmail(request.getEmail());
        user.setPasswordHash(passwordEncoder.encode(request.getPassword()));
        user.setRole(request.getRole() != null ? request.getRole() : "Customer");
        user.setCreatedAt(LocalDateTime.now());
        user.setActive(true);
        userRepository.save(user);

        UserProfile profile = new UserProfile();
        profile.setId(UUID.randomUUID());
        profile.setUserId(user.getId());
        profile.setName(request.getName());
        profile.setAddress(request.getAddress());
        profile.setPhone(request.getPhone());
        userProfileRepository.save(profile);

        return UserResponse.builder()
                .id(user.getId())
                .email(user.getEmail())
                .role(user.getRole())
                .createdAt(user.getCreatedAt())
                .profile(UserResponse.Profile.builder()
                        .name(profile.getName())
                        .address(profile.getAddress())
                        .phone(profile.getPhone())
                        .build())
                .build();
    }
}

File: python-user-management-service/user_service.py
# GH Copilot code - starts
from user import User
from user_profile import UserProfile
from user_registration_request import UserRegistrationRequest
from user_repository import UserRepository
from error_response import ErrorResponse

class UserService:
    def __init__(self, user_repository=None):
        self.user_repository = user_repository or UserRepository(None)

    def register_user(self, registration_data: dict):
        req = UserRegistrationRequest(**registration_data)
        try:
            req.validate()
        except ValueError as e:
            raise ErrorResponse("InvalidInput", str(e))
        existing = self.user_repository.get_user_by_email(req.email)
        if existing:
            raise ErrorResponse("DuplicateEmail", "Email already exists")
        profile = UserProfile(req.name, req.address, req.phone)
        user = User(email=req.email, password=req.password, role=req.role, profile=profile)
        self.user_repository.create_user(user)
        return user
# GH Copilot code - end
